File: desktop/electron/modules/wsl.ts
/**
 * WSL (Windows Subsystem for Linux) support.
 *
 * Ported from src-tauri/src/wsl.rs
 *
 * On Windows, the desktop app can launch and connect to a djinn-server running
 * inside the default WSL distribution. WSL2 shares localhost networking so the
 * server is reachable at 127.0.0.1:{port}.
 *
 * On non-Windows platforms all functions are no-ops / return false.
 */

import { execSync } from "node:child_process";
import { platform } from "node:os";
import { healthCheck } from "./server.js";

// ---------------------------------------------------------------------------
// WSL detection
// ---------------------------------------------------------------------------

/** Check if WSL is available on this machine (Windows only). */
export function isAvailable(): boolean {
  if (platform() !== "win32") {
    return false;
  }
  try {
    const result = execSync("wsl --status", { stdio: "pipe" });
    return result !== null;
  } catch {
    return false;
  }
}

// ---------------------------------------------------------------------------
// WSL daemon management
// ---------------------------------------------------------------------------

/**
 * Ensure a djinn-server daemon is running inside the default WSL distribution.
 *
 * Returns the base URL on success (always `http://127.0.0.1:{port}` because
 * WSL2 shares localhost).
 */
export async function ensureWslDaemon(port: number): Promise<string> {
  if (platform() !== "win32") {
    throw new Error(
      `WSL is not available on this platform (requested port ${port})`,
    );
  }

  // Check if already running.
  try {
    execSync("wsl -e pgrep -f djinn-server", { stdio: "pipe" });
  } catch {
    // Not running -- start it.
    console.log(`Starting djinn-server inside WSL on port ${port}`);
    try {
      execSync(
        `wsl -e sh -c "nohup djinn-server --port ${port} &>/dev/null &"`,
        { stdio: "pipe" },
      );
    } catch (e) {
      throw new Error(`WSL djinn-server start failed: ${e}`);
    }

    // Give it a moment to bind.
    await sleep(500);
  }

  const baseUrl = `http://127.0.0.1:${port}`;

  // Health-check loop.
  for (let i = 0; i < 40; i++) {
    if (await healthCheck(baseUrl)) {
      return baseUrl;
    }
    await sleep(250);
  }

  throw new Error(
    "WSL djinn-server started but health endpoint not reachable",
  );
}

// ---------------------------------------------------------------------------
// Utility
// ---------------------------------------------------------------------------

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: desktop/src-tauri/src/wsl.rs
//! WSL (Windows Subsystem for Linux) support.
//!
//! On Windows, the desktop app can launch and connect to a djinn-server running
//! inside the default WSL distribution. WSL2 shares localhost networking so the
//! server is reachable at `127.0.0.1:{port}`.
//!
//! On non-Windows platforms all functions are no-ops / return `false`.

/// Check if WSL is available on this machine.
#[cfg(target_os = "windows")]
pub fn is_available() -> bool {
    std::process::Command::new("wsl")
        .arg("--status")
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false)
}

#[cfg(not(target_os = "windows"))]
pub fn is_available() -> bool {
    false
}

/// Ensure a djinn-server daemon is running inside the default WSL distribution.
///
/// Returns the base URL on success (always `http://127.0.0.1:{port}` because
/// WSL2 shares localhost).
#[cfg(target_os = "windows")]
pub async fn ensure_wsl_daemon(port: u16) -> Result<String, String> {
    // Check if already running.
    let check = std::process::Command::new("wsl")
        .args(["-e", "pgrep", "-f", "djinn-server"])
        .output()
        .map_err(|e| format!("Failed to run WSL pgrep: {}", e))?;

    if !check.status.success() {
        // Not running — start it.
        log::info!("Starting djinn-server inside WSL on port {port}");
        let start = std::process::Command::new("wsl")
            .args([
                "-e",
                "sh",
                "-c",
                &format!("nohup djinn-server --port {port} &>/dev/null &"),
            ])
            .output()
            .map_err(|e| format!("Failed to start djinn-server in WSL: {}", e))?;

        if !start.status.success() {
            let stderr = String::from_utf8_lossy(&start.stderr).to_string();
            return Err(format!("WSL djinn-server start failed: {}", stderr.trim()));
        }

        // Give it a moment to bind.
        tokio::time::sleep(std::time::Duration::from_millis(500)).await;
    }

    let base_url = format!("http://127.0.0.1:{}", port);

    // Health-check loop.
    for _ in 0..40 {
        if crate::server::health_check(&base_url).await {
            return Ok(base_url);
        }
        tokio::time::sleep(std::time::Duration::from_millis(250)).await;
    }

    Err("WSL djinn-server started but health endpoint not reachable".into())
}

#[cfg(not(target_os = "windows"))]
pub async fn ensure_wsl_daemon(port: u16) -> Result<String, String> {
    Err(format!(
        "WSL is not available on this platform (requested port {port})"
    ))
}
